apollo-federation/src/query_graph/graph_path.rs

```diff
@@ -1686,143 +1657,148 @@ where
-                                                })
-                                                .for_each(|_| {
-                                                    potential_matches.insert(iface.name.clone());
-                                                });
-                                        }
-                                    }
-                                    CompositeTypeDefinitionPosition::Union(union_pos) => {
-                                        if let Ok(un) = union_pos.get(subgraph_schema.schema()) {
-                                            un.members
-                                                .iter()
-                                                .filter(|item| {
-                                                    &item.name == parent_type.type_name()
-                                                })
-                                                .for_each(|_| {
-                                                    potential_matches.insert(un.name.clone());
-                                                });
-                                        }
-                                    }
+                            }
+                            CompositeTypeDefinitionPosition::Interface(
+                                parent_type_in_supergraph,
+                            ) => {
+                                if parent_type_in_supergraph
+                                    .get(supergraph_schema.schema())?
+                                    .implements_interfaces
+                                    .iter()
+                                    .any(|item| &item.name == pos.type_name())
+                                {
+                                    return Ok(true);
                                 }
-                            });
+                            }
+                            _ => {}
+                        }
+                        let pos_in_supergraph: CompositeTypeDefinitionPosition = supergraph_schema
+                            .get_type(pos.type_name().clone())?
+                            .try_into()?;
+                        if let CompositeTypeDefinitionPosition::Union(pos_in_supergraph) =
+                            &pos_in_supergraph
+                        {
+                            if pos_in_supergraph
+                                .get(supergraph_schema.schema())?
+                                .members
+                                .iter()
+                                .any(|item| &item.name == parent_type_in_supergraph.type_name())
+                            {
+                                return Ok(true);
+                            }
+                        }
+                        Ok(false)
+                    })? {
+                        continue;
+                    }
 
-                            // get the selection set for the first match that parses
-                            let selection_set =
```

Comment:
Bugs:
- The selection set should be parsed against the supergraph schema, not the subgraph schema.
- Duck typing semantics means the `@fromContext` selection set may not validate against any of the types individually. The fix here is to parse without validation against the parent type in the supergraph schema, remove any "impossible" type conditions, and then validate.